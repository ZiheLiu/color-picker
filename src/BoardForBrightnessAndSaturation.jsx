"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ReactDOM = require("react-dom");
var add_dom_event_listener_1 = require("add-dom-event-listener");
var const_1 = require("./utils/const");
var Color_1 = require("./utils/Color");
var BoardForBrightnessAndSaturation = /** @class */ (function (_super) {
    __extends(BoardForBrightnessAndSaturation, _super);
    function BoardForBrightnessAndSaturation() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.moveListener = null;
        _this.upListener = null;
        _this.handleMouseDown = function () {
            _this.moveListener = add_dom_event_listener_1.default(window, 'mousemove', _this.handleMouseMove);
            _this.upListener = add_dom_event_listener_1.default(window, 'mouseup', _this.handleMouseUp);
        };
        _this.handleMouseMove = function (e) {
            var x = e.clientX;
            var y = e.clientY;
            _this.pointMoveTo({ x: x, y: y });
        };
        _this.handleMouseUp = function () {
            _this.removeListener();
        };
        _this.removeListener = function () {
            if (_this.moveListener) {
                _this.moveListener.remove();
            }
            if (_this.upListener) {
                _this.upListener.remove();
            }
        };
        _this.pointMoveTo = function (pos) {
            var rect = ReactDOM.findDOMNode(_this).getBoundingClientRect();
            var left = pos.x - rect.left;
            var bottom = rect.bottom - pos.y;
            var _a = _this.props, size = _a.size, color = _a.color, onChange = _a.onChange;
            var maxWidth = rect.width || size;
            var maxHeight = rect.height || size;
            left = Math.min(maxWidth, Math.max(0, left));
            bottom = Math.min(maxHeight, Math.max(0, bottom));
            var newColor = color.copy();
            newColor.saturation = left / maxWidth;
            newColor.brightness = bottom / maxHeight;
            onChange(newColor);
        };
        return _this;
    }
    BoardForBrightnessAndSaturation.prototype.componentWillUnmount = function () {
        this.removeListener();
    };
    BoardForBrightnessAndSaturation.prototype.render = function () {
        var _a = this.props, size = _a.size, prefix = _a.prefix, color = _a.color;
        var hueColor = new Color_1.default({
            h: color.hue,
            s: 1,
            v: 1
        });
        var panelStyle = {
            width: size,
            height: size,
            background: "#" + hueColor.hex
        };
        var panelClass = prefix;
        var saturationClass = prefix + "-saturation-layer";
        var brightnessClass = prefix + "-brightness-layer";
        return (<div style={panelStyle} className={panelClass} onMouseDown={this.handleMouseDown}>
        <div className={saturationClass}/>
        <div className={brightnessClass}/>
      </div>);
    };
    BoardForBrightnessAndSaturation.defaultProps = {
        size: 300,
        prefix: const_1.Prefix + "-panel"
    };
    return BoardForBrightnessAndSaturation;
}(React.Component));
exports.default = BoardForBrightnessAndSaturation;
