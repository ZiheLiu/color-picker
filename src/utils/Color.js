"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var TinyColor = require("tinycolor2");
var Color = /** @class */ (function () {
    // color will be #000 if input is illegal
    function Color(input) {
        this.color = TinyColor('blue');
        this.color = TinyColor(input);
    }
    Color.prototype.copy = function () {
        return new Color(this.color);
    };
    Object.defineProperty(Color.prototype, "saturation", {
        set: function (value) {
            var hsb = this.color.toHsv();
            this.color = TinyColor(__assign({}, hsb, { s: value }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Color.prototype, "brightness", {
        set: function (value) {
            var hsb = this.color.toHsv();
            this.color = TinyColor(__assign({}, hsb, { v: value }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Color.prototype, "hue", {
        get: function () {
            var hsb = this.color.toHsv();
            return hsb.h;
        },
        set: function (value) {
            var hsb = this.color.toHsv();
            this.color = TinyColor(__assign({}, hsb, { h: value }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Color.prototype, "hex", {
        get: function () {
            return this.color.toHex();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Color.prototype, "hexString", {
        get: function () {
            return "#" + this.hex;
        },
        enumerable: true,
        configurable: true
    });
    return Color;
}());
exports.default = Color;
