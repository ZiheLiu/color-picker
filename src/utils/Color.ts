import * as TinyColor from 'tinycolor2';

interface Hsv {
  h: number;
  s: number;
  v: number;
}

class Color {
  color = TinyColor('blue');

  constructor(input: string | Hsv) {
    this.color = TinyColor(input);
  }

  copy() {
    return new Color(this.color);
  }

  set saturation(value: number) {
    const hsb = this.color.toHsv();
    this.color = TinyColor({
      ...hsb,
      s: value
    });
  }

  set brightness(value: number) {
    const hsb = this.color.toHsv();
    this.color = TinyColor({
      ...hsb,
      v: value
    });
  }

  set hue(value: number) {
    const hsb = this.color.toHsv();
    this.color = TinyColor({
      ...hsb,
      h: value
    });
  }

  get hue() {
    const hsb = this.color.toHsv();
    return hsb.h;
  }

  get hex() {
    return this.color.toHex();
  }

  get hexString() {
    return `#${this.hex}`;
  }
}

export default Color;
