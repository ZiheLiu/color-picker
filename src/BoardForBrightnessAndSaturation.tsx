import * as React from 'react';
import * as ReactDOM from 'react-dom';
import addEventListener from 'add-dom-event-listener';

import { Prefix } from './utils/const';
import Color from './utils/Color';

interface PanelProps {
  size?: number;
  prefix?: string;
  color: Color;
  onChange: (color: Color) => void;
}

interface Position {
  x: number;
  y: number;
}

class BoardForBrightnessAndSaturation extends React.Component<PanelProps, {}> {
  static defaultProps = {
    size: 300,
    prefix: `${Prefix}-panel`
  };

  moveListener: any = null;
  upListener: any = null;

  handleMouseDown = () => {
    this.moveListener = addEventListener(
      window,
      'mousemove',
      this.handleMouseMove
    );
    this.upListener = addEventListener(window, 'mouseup', this.handleMouseUp);
  };

  handleMouseMove = (e: any) => {
    const x = e.clientX;
    const y = e.clientY;
    this.pointMoveTo({ x, y });
  };

  handleMouseUp = () => {
    this.removeListener();
  };

  removeListener = () => {
    if (this.moveListener) {
      this.moveListener.remove();
    }
    if (this.upListener) {
      this.upListener.remove();
    }
  };

  pointMoveTo = (pos: Position) => {
    const rect = ReactDOM.findDOMNode(this).getBoundingClientRect();
    let left = pos.x - rect.left;
    let bottom = rect.bottom - pos.y;

    const { size, color, onChange } = this.props;
    const maxWidth = rect.width || size!;
    const maxHeight = rect.height || size!;

    left = Math.min(maxWidth, Math.max(0, left));
    bottom = Math.min(maxHeight, Math.max(0, bottom));

    const newColor = color.copy();
    newColor.saturation = left / maxWidth;
    newColor.brightness = bottom / maxHeight;

    onChange(newColor);
  };

  componentWillUnmount() {
    this.removeListener();
  }

  render() {
    const { size, prefix, color } = this.props;

    const hueColor = new Color({
      h: color.hue,
      s: 1,
      v: 1
    });

    const panelStyle = {
      width: size,
      height: size,
      background: `#${hueColor.hex}`
    };
    const panelClass = prefix;
    const saturationClass = `${prefix}-saturation-layer`;
    const brightnessClass = `${prefix}-brightness-layer`;

    return (
      <div
        style={panelStyle}
        className={panelClass}
        onMouseDown={this.handleMouseDown}
      >
        <div className={saturationClass} />
        <div className={brightnessClass} />
      </div>
    );
  }
}

export default BoardForBrightnessAndSaturation;
