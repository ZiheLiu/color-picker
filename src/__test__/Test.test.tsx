import * as React from 'react';
import * as renderer from 'react-test-renderer';

import BoardForBrightnessAndSaturation from '../BoardForBrightnessAndSaturation';

import Color from '../utils/Color';

test('Text of Test should be correct', () => {
  const color = new Color({
    h: 0,
    s: 1,
    v: 1
  });
  const onChange = (c: Color) => {
    color = c;
  };
  const component = renderer.create(
    <BoardForBrightnessAndSaturation color={color} onChange={onChange} />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
