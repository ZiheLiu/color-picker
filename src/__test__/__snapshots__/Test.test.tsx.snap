// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text of Test should be correct 1`] = `
<div
  className="cp-panel"
  onMouseDown={[Function]}
  style={
    Object {
      "background": "#ff0000",
      "height": 300,
      "width": 300,
    }
  }
>
  <div
    className="cp-panel-saturation-layer"
  />
  <div
    className="cp-panel-brightness-layer"
  />
</div>
`;
